### DISCUSSION TITLE
API request from Nextjs app returns 42703 error but Postman request works as expected

### DISCUSSION THREAD

**Discussion by kavitang** [supabase/supabase#17954] [Questions]
Hey,

I'm trying to set up the `null` Postgrest 11 filtering https://postgrest.org/en/stable/references/api/resource_embedding.html#null-filtering-on-embedded-resources, to fetch records that don't exist in a join table. 

In my DB I have `table1` and `table2` with a join table `table1_table2` which has a composite primary key pointing to `table1` and `table2`. I'm trying to fetch all records in `table1` without an entry in `table1_table2`. As per Postgrest 11 I can do this with `/table1?select=table1_table2(table1_id,table2_id)id,other_fields&table1_table2=is.null`. 

When I tried this, it works perfectly fine in Postman, but when I do the same through the Nextjs application, it is returning a 42703 error stating that `column "table1_table1_table2_1" does not exist`. I even copied the payload as is from the Nextjs request, and pasted it in Postman to be sure, and it works perfectly.

I'm trying this on a new DB so it is using Postgrest 11, and since it works in Postman, I'm not sure why the exact same request from the app is returning this error. I have noticed this discrepancy in responses for other requests as well, but can't find a reason for this as the requests are the exact same, going to the same endpoints.

Any insights here would help as this is a feature I need to implement for the app, and I'm not able to figure out why it would work for one request but not the other.

Thanks

**Comment by GaryAustin1** [comment]
Can you show your next.js API call?

**Reply by GaryAustin1** [reply]
Also how are you setting the JWT in the authorization header for Postman?  Could you be having an RLS issue on the join table?
The column name certainly seems strange in: `column "table1_table1_table2_1" does not exist`.  Where did the _1 come from and the table1_table1 part is odd.

**Reply by kavitang** [reply]
Sure this is the request sent from the Nextjs app, copy pasted into Postman where it works, but here it returns the error

`https://<supabaseurl>/rest/v1/table1?select=table1_table2%28id%2Cnickname%2Ctable1_id%2Ctable2_id%2Ctable2_id%28id%2Cicon%2Cname%29%29%2Ccreated_at%2Cupdated_at&order=id.desc.nullslast&is_active=eq.true&is_test=eq.false&table1_table2=is.null&offset=0&limit=200`

Apikey etc is sent correctly, the error returned is

```
{
    "code": "42703",
    "details": null,
    "hint": null,
    "message": "column \"table1_table1_table2_1\" does not exist"
}
```

Structure in DB is 
`table1` - PK `id`
`table2` - PK `id`
`table1_table2` - PK `(table1_id, table2_id, id)`  `table1_id` and `table2_id` are foreign keys

**Reply by kavitang** [reply]
Authorization Header is set correctly in Postman, the request works fine, I can see the expected results. In the Nextjs app, the authorization header etc is set correctly. I don't think this is an RLS issue, as Postman request works fine, and the error returned by the Nextjs app is not related. The one difference I noticed is that here `table1_table2` has `id` in it's composite key, but when I try the same request with a join table that only has foreign keys in the composite key, the error returned changes slightly, it's still 42703, but the message is `column "table1.another_join_table_to_table1" does not exist`. 

In this case the request was `/table1?select=another_join_table_to_table1(),id,other_fields&another_join_table_to_table1=is.null`

**Reply by kavitang** [reply]
Ok I created a Sandbox to see if I can replicate this issue, and it seems to be working perfectly fine there in both Postman and the Nextjs app. So there is something in my app that is causing this problem...Any idea what could affect the request? What could cause the DB to not recognise the join table from only the app?

The requests are identical, I copy pasted the payload, but the Next app seems to be returning a different response..

 The one difference here is the use of supabase-auth-nextjs package in the Nextjs app that is not working, where the Sandbox app is directly using the Supabase js package... I'll check to see if that is the issue..

**Reply by GaryAustin1** [reply]
Are you sure you are meeting RLS for your extra tables in next.js.  You might try and select one by itself.

**Reply by kavitang** [reply]
`table1` is the only table with an RLS policy that allows all operations for this user. `table2` and `table1_table2` have no RLS policies enabled. Also would an RLS issue not cause the same response through Postman, where I'm passing the exact same JWT in the Authorization header? I tried adding the supabase-auth-nextjs package to the sandbox and it still works, only my main Next js app is returning this result, which is strange as the API request is going to the same end point with the same data, but returns different results.. is there any way it could be using the older Postgrest version through the app? This was not possible in Postgrest 10, which is why we were waiting for Postgrest 11 to implement the null filtering...

**Reply by GaryAustin1** [reply]
I assume you mean you have RLS disabled on those other tables versus enabled with no policies.
If you actually looked at the jwt in the authorization header of your next.js call and it was the user jwt then not likely involved.  If you just assume you are signed in when you make the call, that often turns out to be a bad assumption.

**Reply by GaryAustin1** [reply]
"table1_table1_table2_1"  I still don't know how that could be generated in that format with columns and tables names you have shown.  Where would table2_1 in particular come from?  You have not shown your actual next.js call either.

**Reply by kavitang** [reply]
Yes I meant RLS is disabled for the other two tables. And also yes I copied the Authorization header from the Next js call, and used that in the Postman request as well. The main table being queried `table1` uses RLS and allows authenticated users only, so the Postman request would not have worked without the Auth header. Also, other requests from the Next js app for data from these tables are all working fine, it's only the null filtering which is returning this response.

**Reply by kavitang** [reply]
The call is very simple `https://projectid.supabase.co/rest/v1/table1?select=id%2Cname%2Ctable1_table2%28table1_id%2Ctable2_id%29&table1_table2=is.null` 

Payload below: 

```
select: id,name,table1_table2(table1_id,table2_id)
table1_table2: is.null
```
There is no reference to `table1_table1_table2_1` in the request in any form, so I'm also not sure where this is coming from. Like I mentioned in an earlier comment, if I try this same null filtering with `table1` and another join table without `id` in the composite key, the error message is `column "table1.another_join_table" does not exist`. Not sure why it thinks the join table name is a column on table1 instead of treating it as a join table.. The actual app is complicated but the request being sent is still identical to the Postman request, I'm really stumped by how the app is playing a role in the API response being different, when the payload is the same and is going to the same endpoint as the Postman request..not sure what I'm missing here... 😅

**Reply by GaryAustin1** [reply]
Are you generating your own REST API request in next.js or using the javascript client?

Since you can see the request this is probably not relevant, but many users with next.js 13 are having caching in next.js on url's mess up their debugging when just changing the Supabase side (RLS, data in table, etc).  The cache latches on to the first request and its error and then no matter the changes made to the database side, none of those are reflected.

I don't have much else at the moment.

**Reply by kavitang** [reply]
I am using the javascript client, but yes, the final request is still to the same DB that works fine through Postman, and has the same payload. I'll try to simplify the app to see if I'm able to identify which layer is causing a problem. Thanks for the suggestions and the help! I'll update once I find something 🤞

**Comment by kavitang** [comment]
@GaryAustin1 - I found the cause, also reproducible in this sandbox. If you add `count: estimated/exact` to the select, or pass `Prefer: count=*` as a header to the request it shows this error when doing null filtering. It works absolutely fine without this header only. 
[Sandbox](https://codesandbox.io/p/sandbox/bold-hooks-fqqkzc?file=%2Futils%2Fapi.ts%3A14%2C19). I'm assuming this is a bug.

```
// Returns the error

const { data } = await supabaseClient
    .from("table1")
    .select("id,table1_table2(table1_id,table2_id)", {
      count: "estimated" / "exact",
      head: false,
    })
    .is("table1_table2", null);

// Works correctly
const { data } = await supabaseClient
    .from("table1")
    .select("id,table1_table2(table1_id,table2_id)")
    .is("table1_table2", null);
```
